File: core/evidence/recursive_lineage_engine.py
from __future__ import annotations

from typing import Any, Dict, List


def preserve_recursive_lineage(
    lineage: Dict[str, Any],
    evidence: List[str],
    ambiguities: List[str],
    uncertainties: List[str],
    contradicted: Dict[str, Any],
) -> Dict[str, Any]:
    stages = lineage.get("stages", []) if isinstance(lineage, dict) else []
    depth = len(stages) if isinstance(stages, list) else int(lineage.get("depth", 0) or 0)
    pairs = contradicted.get("pairs", []) if isinstance(contradicted, dict) else []
    return {
        "depth": depth,
        "evidence_ancestry": list(evidence),
        "ambiguity_ancestry": list(ambiguities),
        "uncertainty_ancestry": list(uncertainties),
        "contradiction_ancestry": list(pairs),
        "entropy_ancestry_preserved": True,
        "instability_ancestry_preserved": True,
        "decay_prevented": True,
    }

File: src/evidence/recursiveLineageEngine.ts
/**
 * Converted from Python: core/evidence/recursive_lineage_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function preserveRecursiveLineage(lineage: any, evidence: any, ambiguities: any, uncertainties: any, contradicted: any): any {
  var stages: any = (((lineage !== null && typeof lineage === "object" && !Array.isArray(lineage) && !(lineage instanceof Set) && !(lineage instanceof Map))) ? py.get(lineage, "stages", []) : []);
  var depth: any = ((Array.isArray(stages)) ? py.len(stages) : py.toInt(py.or2(py.get(lineage, "depth", 0), () => (0))));
  var pairs: any = (((contradicted !== null && typeof contradicted === "object" && !Array.isArray(contradicted) && !(contradicted instanceof Set) && !(contradicted instanceof Map))) ? py.get(contradicted, "pairs", []) : []);
  return {"depth": depth, "evidence_ancestry": [...py.iter(evidence)], "ambiguity_ancestry": [...py.iter(ambiguities)], "uncertainty_ancestry": [...py.iter(uncertainties)], "contradiction_ancestry": [...py.iter(pairs)], "entropy_ancestry_preserved": true, "instability_ancestry_preserved": true, "decay_prevented": true};
}
